--- miniaturo-0.5.0/Cargo.toml	2022-11-05T14:06:58+00:00
+++ miniaturo-0.5.0/Cargo.toml	2023-04-30T12:28:44.419763+00:00
@@ -10,7 +10,7 @@
 anyhow = "1"
 clap = { version = "4", features = ["std", "derive", "help", "usage", "cargo", "error-context", "wrap_help"], default-features = false }
 image = { version = "0.24", features = ["png", "jpeg", "tiff"], default-features = false }
-libopenraw-rs = { git = "https://gitlab.freedesktop.org/libopenraw/libopenraw-rs.git", rev = "f992e4ffd132dde852226bd13af1ea6d0a15e42e" }
+libopenraw = "0.1.2"
 
 [dev-dependencies]
 assert_cmd = "2"
